### DISCUSSION TITLE
2025 Outlook for Q1-Q2

### DISCUSSION THREAD

**Discussion by ayerofieiev-tt** [tenstorrent/pytorch2.0_ttnn#724] [Announcements]
# 2025 Outlook for Q1-Q2 
Lets quickly recap second half of 2024
* Got a basic test suite for models getting from 0 to ~150 variations
* Got basic CI running
* Got from 0 models running to 66 models running within TT-NN
* Most importantly, we collected a ton of data for the team to improve the generality of many TT-NN operations

In 2025 the priority remains to grow the number of models we can run. During 2025 Q1-Q2 focus shifts towards performance and accuracy.

There are 5 high level projects 
1. Release Wheel (Installation & Distribution)
2. Enable run of models in Data Parallel mode
3. Expand Model Coverage
4. Optimize performance ⭐️
5. Enable run of larger models in Tensor Parallel mode

## 1. Release Wheel (Installation & Distribution)

**Objective:** Ensure the TT-NN compiler is easily installable from GitHub releases and PyPI.

**Key Results:**

* **M1:** Regularly produce and upload wheels to GitHub Releases, ensuring `pip install <gh-artifact-link>` works reliably.  
* **M2:** Publish wheels to PyPI, enabling `pip install pytorch-ttnn` for end-users.

Acceptance criteria:
* User must manually setup machine (default Tenstorrent setup, like tt-flash, tt-smi, hugepages, etc)  
* User does not have to manually install TT-NN  
* User can install pytorch-ttnn wheel from pypi  

Process:
* Develop packaging scripts and a CI job that runs basic tests on a produced wheel and publishes it on Github  
* Extend the script to upload releases to PyPi  
* Maintain packaging scripts for the duration of this SoW

## 2. Enable run of models in Data Parallel mode

**Objective:** Enable users to increase throughput for parallel workloads by running them on multidevice systems.

Notes:

* Expected workflow: user sets compiler into “Data Parallel” mode. Compiler just does it. Supported by TT-NN already.   
* Switch to MeshDevice  
* Weights must replicated  
  `ttnn_tensor = from_torch(torch_tensor,... mesh_mapper=ReplicateTensorToMesh(mesh_device), device=mesh_device)`  
* Inputs must be sharded across the Batch dimension (dim=0)  
  `ttnn_tensor = from_torch(torch_tensor,... mesh_mapper=ShardTensorToMesh(mesh_device, dim=0), device=mesh_device)`  
* To collect results back  
  `torch_tensor = to_torch(torch_tensor,... mesh_composer=ConcatMeshToTensor(mesh_device, dim=0), device=mesh_device)`  
* Same flow can be used for DataParallel and Single device (assume MeshDevice with a single device open)  
* Don’t need to care about CCL

## 3. Expand Model Coverage

**Objective:** Increase end-to-end device execution to 100 out of 153-model corpus

Notes:
* We continue to prioritize models according to the clustered list
* Need to refactor `to_tt_pass.py` to better organize existing lowering to make it easier to locate and inspect

## 4. Optimize performance ⭐️

**Objective:** Provide strong out-of-the-box performance so users see tangible benefits immediately, and establish clear benchmarks against a reference implementation.

Performance Optimization Techniques:

* Keep weights on device  
  * [https://github.com/bdhirsh/pytorch\_open\_registration\_example/blob/master/cpp\_extensions/open\_registration\_extension.cpp](https://github.com/bdhirsh/pytorch_open_registration_example/blob/master/cpp_extensions/open_registration_extension.cpp)  
  * [https://github.com/pytorch/pytorch/pull/80477/files\#diff-ab5a74c358ad7af7a6fb241e7655670f15e9dbea223e724eba18033321e25b52](https://github.com/pytorch/pytorch/pull/80477/files#diff-ab5a74c358ad7af7a6fb241e7655670f15e9dbea223e724eba18033321e25b52)  
* Minimize TMs between OPs (pad, unpad, tilize, untilize, reshape)  
* Maximize intermediate tensors in L1 (minimize spill to DRAM)  
* Choose best sharding for each OP  
* Maximize length of “sharding chain” (another way of saying minimize TMs)  
* Pattern match some OPs into some other performant OPs (softmax, concatenated heads)  
* Pattern match the whole model. 

## 5. Enable run of larger models in Tensor Parallel mode

**Objective:** Enable seamless scaling of large models (e.g., Llama 70B) across multiple devices

Notes:

* Expected workflow: user sets compiler into “Tensor Parallel” mode. Compiler uses it’s “sharding optimization heuristic” from the performance optimization goal to shard across multiple devices.  
* Similarly to Data Parallel, tensors have to be replicated/sharded across devices, same API can be used for it, but different strategies are required. E.g. instead of replicating weights \- they now have to be sharded across the mesh of devices  
* CCL operations have to be injected to collect a result from multiple devices  
* [Programming Mesh of Devices with TT-NN.md](https://github.com/tenstorrent/tt-metal/blob/main/tech_reports/Programming%20Mesh%20of%20Devices/Programming%20Mesh%20of%20Devices%20with%20TT-NN.md)

## Explorations
We are exploring a couple other topics:
* Integration as a native PyTorch device
* PyTorch Inductor
* TT-NN JIT
